fix(appium,support): fix installation problems

boneskull opened the pull request:
Ref: #17073

this change uses `--legacy-peer-deps` for newer versions of npm to
avoid installation of appium when installing extensions in `~/.appium`.  this does not work if the extension contains a shrinkwrap with appium in it, though.
in that case, an extra copy of appium will always be installed.

it creates a symlink, e.g., `~/.appium/node_modules/appium` to the currently-running appium, wherever that is.  could not abuse `NODE_PATH` to do this, as changing it on-the-fly has no effect (it must be set at start of `node` process)

this is rather naive and does not check that the symlink is actually the same as the running appium if it already exists. that might cause some weird problems with multiple installs of appium on the same machine, but if it does, we can fix them as they arise.

boneskull committed:
fix(appium,support): fix installation problems
Ref: #17073

this change uses --legacy-peer-deps for newer versions of npm to
avoid installation of appium when installing extensions in ~/.appium.  this does not work if the extension contains a shrinkwrap with appium in it, though.
in that case, an extra copy of appium will always be installed.

it creates a symlink, e.g., ~/.appium/node_modules/appium to the currently-running appium, wherever that is.  could not abuse NODE_PATH to do this.

this is rather naive and does not check that the symlink is actually the same as the running appium if it already exists. that might cause some weird problems with multiple installs of appium on the same machine, but if it does, we can fix them as they arise.

jlipps reviewed (approved):
ok let's give it a shot!

jlipps commented:
hmm seems like build is unhappy with this

mykola-mokhnach commented on packages/appium/lib/main.js:
```diff
@@ -170,6 +172,18 @@ function areServerCommandArgs(args) {
+  // of `appium`, which lives wherever it lives (see `rootDir`).
+  if (!(await hasAppiumDependency(appiumHome))) {
+    try {
+      await fs.symlink(rootDir, path.join(appiumHome, 'node_modules', 'appium'));
```
would this work on Windows?

mykola-mokhnach replied:
I assume we need to use junction there

mykola-mokhnach replied:
I assume we need to call mkdirp first to create `node_modules` root in `appiumHome`

boneskull replied:
are you sure we need `junction`?  I thought that was a hard link--e.g., if `~/.appium/node_modules/appium` is deleted then e.g., `$NPM_PREFIX/node_modules/appium` also gets deleted?

mykola-mokhnach replied:
on Windows - yes. it cannot create links for folders. Check https://github.com/nodejs/node/issues/18518, but you could try it yourself. Btw I know a company, which provides a whole VM cloud, where Windows-related stuff could also be tested ;)

mykola-mokhnach commented on packages/appium/lib/main.js:
```diff
@@ -170,6 +172,18 @@ function areServerCommandArgs(args) {
+  if (!(await hasAppiumDependency(appiumHome))) {
+    try {
+      await fs.symlink(rootDir, path.join(appiumHome, 'node_modules', 'appium'));
+    } catch (err) {
```
is rootDir pointing to the root of the global Appium installation?
Maybe then it would make sense to put a symlink to `$HOME/.node_modules` according to https://nodejs.org/api/modules.html#loading-from-the-global-folders

boneskull commented:
> I assume we need to call mkdirp first to create `node_modules` root in `appiumHome` 

right, I think so.

I am not sure about windows.  I think we need to expand our CI matrix.

boneskull commented:
> Maybe it would make sense to put a symlink to `$HOME/.node_modules` according to https://nodejs.org/api/modules.html#loading-from-the-global-folders 

hmm, yes, that may be a way forward. I'll investigate

mykola-mokhnach reviewed (approved):
LGTM, however I'd still check this functionality on Windows before merging just to be on the safe side

jlipps commented:
Am trying to check on Windows, however it seems there are some other problems on Windows potentially so it's not straightforward. Really need to get Windows in our CI I guess!

boneskull commented:
Ok, I tested this on Windows 10.  This is what I did:

1. Install [nvm-windows](/coreybutler/nvm-windows)
2. Open PowerShell as admin
3. `nvm install latest` (18.x)
4. `nvm use latest`
5. clone this repo
6. `npm install` (build appium with Node v18.x and npm v8.x) **this updates `package-lock.json` -- it needs to be added to this PR
7. `nvm install 14` (comes w/ npm v6.x)
8. `nvm use 14.19.3`
9. create some new tmp dir and navigate to it.  I did this just to be safe, but it's unclear if skipping this step would cause a problem
10. `node \path\to\packages\appium\build\lib\main.js --show-config` works and creates a link to `appium` working copy (subpath `./packages/appium`) in `~/.appium/node_modules`
11. `node \path\to\packages\appium\build\lib\main.js driver install xcuitest` succeeds, but I see a warning about a `test` driver. This is a bug and needs further investigation (but likely only affects dev envs)
13. Running `node \path\to\packages\appium\build\lib\main.js` shows another error about the "test" driver and refuses to load it but otherwise does not complain about xcuitest and loads it successfully
14. Uninstallation of `xcuitest` works.  However, it seems to blow away the symlink.   This may be avoidable by adding an `appium` dependency in `~/.appium/package.json` pointing back to the install path, but may not be an issue at all--the symlink is restored upon subsequent runs. I want to double-check this doesn't just happen at installation of an extension.
15. `nvm use latest` and repeating the `xcuitest` install and running of the server is identical to the output when using Node v14.x/npm v6.x.

So I guess I'm not entirely happy with it, but it seems to be good enough.  I ran into problems at some point due to the `package-lock.json` mismatch which caused really weird errors (@jlipps also experienced this, it sounds like).  

Windows needs to be in our build matrix and that should be high priority.

jlipps commented:
Nice! I'm fine to merge this then when the requisite action items are complete

boneskull commented:
oh i also saw that warning about fake-driver being a dep **and** a devdep. I'll see if that's real or not.

jlipps commented on packages/appium/lib/main.js:
```diff
@@ -170,6 +172,19 @@ function areServerCommandArgs(args) {
+  if (!(await hasAppiumDependency(appiumHome))) {
+    try {
+      await fs.mkdirp(path.join(appiumHome, 'node_modules'));
+      await fs.symlink(rootDir, path.join(appiumHome, 'node_modules', 'appium'), 'junction');
```
woah what is this `junction` business?

jlipps replied:
oh sorry i missed the `)` and misread `junction` as the last argument of `path.join`, not an arg of `fs.symlink`

boneskull commented:
got sidetracked, looking at this test driver thing now

boneskull committed:
fix(support): if appium is installed extraneously, do not consider it a dependency

This can happen if someone does something like `npm install --no-save appium`, _or_ if npm changes its shrinkwrap behavior for reasons that cause `appium` to be installed in `$APPIUM_HOME/node_modules` _(cough)_

boneskull commented:
The "test driver" thing was a bug. I added a fix/test.

boneskull commented:
I should probably explain:

The symlink to `appium` was tricking `@appium/support` into thinking that--since `node_modules/appium` lives under `$APPIUM_HOME`--Appium was installed as a dependency. It checks for this by doing an `npm list appium` in `$APPIUM_HOME`, and sure enough, there it is.  However, since it's not in `$APPIUM_HOME/package.json`, it's considered "extraneous" by npm. The fix was return `false` if that flag is present in the output of the `npm` command.

boneskull commented:
[This is cute](https://github.com/appium/appium/runs/7256786600?check_suite_focus=true#step:7:13179)

boneskull commented:
The `--legacy-peer-deps` flag broke the install of `uiautomator2` with npm v7+ because our `appium` symlink gets blown away during the `npm install uiautomator2`. Some portion of this behavior depends on the contents of the shrinkwrap file, which has recently changed.
